### DISCUSSION TITLE
Cara Deploy ke Shared Hosting

### DISCUSSION THREAD

**Discussion by ahmdakbrhbb** [realodix/urlhub#929] [General]
halo mas saya kan mau coba menggunakan repo mas nya, dan saya sudah coba di localhost memang berjalan seperti biasanya dan menggunakan version terbaru dari php dan composer, namun untuk cara upload ke hosting cpanel itu bagaimana ya?

**Comment by ahmdakbrhbb** [comment]
apakah ada solusi

**Comment by realodix** [comment]
Halo, info yang Anda sebutkan kurang lengkap. Agar Saya tidak kebanyakan menebak, Saya butuh info lebih lanjut.

> saya sudah coba di localhost memang berjalan seperti biasanya dan menggunakan version terbaru dari php dan composer

Masalah extensi GMP di PHP 8.1 yang ada di https://github.com/realodix/urlhub/discussions/745#discussioncomment-7535030 masih relevan? Kalau lihat kalimat Anda di atas, sepertinya sudah tidak relevan.

> namun untuk cara upload ke hosting cpanel itu bagaimana ya?

Ini maksudnya 
- Cara pasang UrlHub di server shared hosting melalui cPanel?
- atau cara pasang di server VPS (atau di atasnya) melalui cPanel?

**Reply by ahmdakbrhbb** [reply]
baik mas, saya mau menanyakan saja bagaimana cara pasang UrlHub di server shared hosting melalui cpanel, karena saya saat ingin pasang selalu tidak bisa di karenakan extension gmp tidak ada di php 8.1 mas

**Reply by ahmdakbrhbb** [reply]
kemungkinan jika ada tutorial atau solusinya boleh beri tahu ya mas

**Reply by ahmdakbrhbb** [reply]
dan apakah langsung pasang saja upload file nya ke shared hosting cpanel saya saja, tanpa perlu untuk install asset lainnya atau sebagainya? karena sudah saya ikutin semua langkah itu di local dan saya jadikan zip

**Comment by realodix** [comment]
> selalu tidak bisa di karenakan extension gmp tidak ada di php 8.1 mas

Berdasarkan dokumentasi PHP, ekstensi GMP sudah ada dan aktif secara default.

**Solusi 1**
Jika lingkungan PHP Anda tidak aktif GMPnya, Anda hanya perlu mengaktifkannya saja.

**Solusi 2**
Jika solusi 1 tidak memungkinkan. Ekstensi GMP digunakan untuk mengantisipasi jika seseorang mengganti settingan default UrlHub pada https://github.com/realodix/urlhub/blob/744f4b5a9c6a851da35a8129a8fd2c0f07cd7168/.env.example#L9 ke angka yang cukup besar (contoh angka 15) dan ini hanya digunakan pada halaman about saja.

Settingan default UrlHub angkanya cukup kecil, Anda bisa dengan mudah merubah kode ini

https://github.com/realodix/urlhub/blob/744f4b5a9c6a851da35a8129a8fd2c0f07cd7168/app/Services/KeyGeneratorService.php#L108

jadi `return (int) pow($nChar, $strLen);`

Lalu hapus https://github.com/realodix/urlhub/blob/744f4b5a9c6a851da35a8129a8fd2c0f07cd7168/composer.json#L9

Jika halaman about hampir tidak pernah dibuka, langsung hapus saja `"ext-gmp": "*"` pada file composer.json

---

> bagaimana cara pasang UrlHub di server shared hosting melalui cpanel
> dan apakah langsung pasang saja upload file nya ke shared hosting cpanel saya saja, tanpa perlu untuk install asset lainnya atau sebagainya? karena sudah saya ikutin semua langkah itu di local dan saya jadikan zip

UrlHub berbasis Laravel dan sayangnya Laravel tidak didesain untuk shared hosting. Perlu sedikit tricky untuk melakukannya.

Jika sudah selesai install dan berjalan dengan lancar, zip aja lalu upload. Setelah itu lakukan penyesuaian di beberapa file.

Ada cukup banyak artikel dan video di YouTube yang membahas cara deploy laravel di shared hosting. Semoga artikel ini bisa cukup memabantu https://www.iankumu.com/blog/deploy-laravel-on-shared-hosting/

**Reply by ahmdakbrhbb** [reply]
setelah semua di setting apa saja mas yg di konfigurasi?

**Reply by realodix** [reply]
Ini konfigurasi untuk apa ya? Jika sudah sampai di tahap selesai (UrlHub sudah run), semua konfiguarasi default sudah diterapkan. Anda dapat menyesuaikan semua konfigurasi sesuai selera.

**Comment by ahmdakbrhbb** [comment]
![Cuplikan layar 2023-11-12 000256](https://github.com/realodix/urlhub/assets/101414231/2920f890-a9b8-4735-b85a-46f0152f8dc7)
mas untuk masalah sekarang seperti ini, ada solusi tidak ya mas? karena saat saya npm di cpanel terminal nya itu tidak bisa

**Reply by realodix** [reply]
Itu karena UInya belum Anda buat. Lanjutkan langkah ini https://github.com/realodix/urlhub?tab=readme-ov-file#compiling-assets-with-laravel-mix.

**Reply by ahmdakbrhbb** [reply]
iyaa mas namun, saya mau npm install di terminal shared hosting cpanel saya tidak bisa selalu npm: command not found

**Reply by ahmdakbrhbb** [reply]
namun di cpanel ini support software nodejsnya, apa harus di pisah atau upload file nya juga ke situ ya?

**Reply by ahmdakbrhbb** [reply]
![Cuplikan layar 2023-11-12 003944](https://github.com/realodix/urlhub/assets/101414231/6c9ca970-7409-4c64-9c67-cf97f05a0321)
seperti ini mas terminal di cpanel nya

**Reply by realodix** [reply]
Karena ini shared hosting, ada kemungkinan memang Anda tidak bisa melakukannya semuanya langsung di server.

> saya sudah coba di localhost memang berjalan seperti biasanya

Solusinya semua langkah yang ada di readme Anda lakukan di komputer Anda terlebih dahulu seperti yang Anda katakan. Setelah semuanya beres, lalu zip dan upload ke server. Lalu lakukan sedikt modifikasi https://www.iankumu.com/blog/deploy-laravel-on-shared-hosting/

Kalau Saya lebih suka melakukan modifikasinya di komputer, jadi tinggal upload saja dan setting database.

**Reply by ahmdakbrhbb** [reply]
![image](https://github.com/realodix/urlhub/assets/101414231/e510634d-e4e7-4c25-92a2-a20b320cf112)
mas mau tanya app startup file nya yg buat di npm install tuh yg apa ya mas

**Reply by ahmdakbrhbb** [reply]
saya sudah npm install semua ikuti readme saat saya upload ke hosting saya selalu begitu

**Reply by ahmdakbrhbb** [reply]
soalnya file public saya dengan file yg lainya saya pisah mengikuti tutorial

**Reply by ahmdakbrhbb** [reply]
sl.rt06.my.id coba mas cek domain saya ini saya subdomain soalnya

**Comment by realodix** [comment]
> mas mau tanya app startup file nya yg buat di npm install tuh yg apa ya mas
> https://github.com/realodix/urlhub/discussions/929#discussioncomment-7543883

Waduh, jadi melebar kemana-mana ya? Ini gak perlu sama sekali.

> https://sl.rt06.my.id/

Ada error Mix manifest not found at: `sl-rt06/public/mix-manifest.json`. Ini karena `mix-manifest.json` tidak ada.

Setelah [realodix/urlhub/readme.md#installation-instructions](https://github.com/realodix/urlhub?tab=readme-ov-file#installation-instructions) lalu ikuti [realodix/urlhub/readme.md#using-npm](https://github.com/realodix/urlhub?tab=readme-ov-file#using-npm). Langkah no. 2 akan menghasilkan file `mix-manifest.json` serta file CSS dan Javascript pada folder public.

<details>
  <summary>Screenshot langkah no. 2</summary>
  
  ![image](https://github.com/realodix/urlhub/assets/1314456/fb3b7ccd-77c1-49cb-8f73-403f4b1e02b6)
</details>

Sampai sini sudah selesai, zip dan upload. Lalu lakukan beberapa penyesuaian untuk shared hosting.

Mungkin video ini dapat membantu https://www.youtube.com/watch?v=VYz7x7fv5jc

---

Sebagai catatan, folder `node_modules` tidak perlu ikutan diupload karena sudah tidak diperlukan lagi. Lumayan besar juga itu bisa ratusan MB.

<details>
  <summary>Screenshot</summary>
  
  ![image](https://github.com/realodix/urlhub/assets/1314456/fa8c083d-6448-4818-b6b7-e733ac8ae011)
</details>

**Reply by ahmdakbrhbb** [reply]
pesan error tetap sama padahal saya sudah mengikuti cara mas nya, dan selalu mainfest gada sedangkan file tersebut ada di file public nya

**Reply by ahmdakbrhbb** [reply]
![image](https://github.com/realodix/urlhub/assets/101414231/45a4a1c3-bd85-4e98-b4f0-7a35e8cf7523)
tuh mas sudah ada file semua yg saat di npm install

**Reply by ahmdakbrhbb** [reply]
owh sudah mas aman

**Reply by ahmdakbrhbb** [reply]
![image](https://github.com/realodix/urlhub/assets/101414231/4f026d98-6839-42ea-ac9c-b7f86fcaab1b)
nah mas error begitu itu dikarenakan apa ya?

**Reply by realodix** [reply]
Itu terjadi ketika unique string generator sudah mencapai batas maksimum. Ada yang dirubah ya? Settingan default cukup besar angka maksimumnya

![image](https://github.com/realodix/urlhub/assets/1314456/79b3c2b4-42e7-42c9-8f45-a3f163e02164)

**Comment by realodix** [comment]
Halo, ada kemungkinan bug yang menyebabkan user yang tidak login (guest) gagal untuk menyingkat link.

Jika itu terjadi di Anda, Saya sudah menerapkan perbaikan kecil https://github.com/realodix/urlhub/commit/a2266c0b68fd201d3d6e09ab6778014345692b07
